src/gui/torrentoptionsdialog.cpp

```diff
@@ -0,0 +1,366 @@
+    connect(m_ui->spinDownloadLimit, qOverload<int>(&QSpinBox::valueChanged)
+            , this, [this](const int value) { updateSliderValue(m_ui->sliderDownloadLimit, value); });
+
+    connect(m_ui->checkMaxRatio, &QCheckBox::toggled, this
+            , [this]() { m_ui->spinRatioLimit->setEnabled(m_ui->checkMaxRatio->isChecked()); });
+    connect(m_ui->checkMaxTime, &QCheckBox::toggled, this
+            , [this]() { m_ui->spinTimeLimit->setEnabled(m_ui->checkMaxTime->isChecked()); });
+
+#if (QT_VERSION >= QT_VERSION_CHECK(5, 15, 0))
+    connect(m_ui->buttonGroup, &QButtonGroup::idClicked, this, &TorrentOptionsDialog::handleRatioTypeChanged);
+#else
+    connect(m_ui->buttonGroup, qOverload<int>(&QButtonGroup::buttonClicked)
+            , this, &TorrentOptionsDialog::handleRatioTypeChanged);
+#endif
+
+    Utils::Gui::resize(this);
+}
+
+TorrentOptionsDialog::~TorrentOptionsDialog()
+{
+    delete m_ui;
+}
+
+void TorrentOptionsDialog::accept()
+{
+    if (m_ui->radioTorrentLimit->isChecked() && !m_ui->checkMaxRatio->isChecked() && !m_ui->checkMaxTime->isChecked())
+    {
+        QMessageBox::critical(this, tr("No share limit method selected"), tr("Please select a limit method first"));
+        return;
+    }
+
+    const auto *session = BitTorrent::Session::instance();
+    for (const BitTorrent::InfoHash &hash : asConst(m_torrentHashes))
+    {
+        BitTorrent::TorrentHandle *torrent = session->findTorrent(hash);
+        if (!torrent) continue;
+
+        if (m_initialValues.upSpeedLimit != m_ui->spinUploadLimit->value())
+            torrent->setUploadLimit(m_ui->spinUploadLimit->value() * 1024);
+        if (m_initialValues.downSpeedLimit != m_ui->spinDownloadLimit->value())
+            torrent->setDownloadLimit(m_ui->spinDownloadLimit->value() * 1024);
+
+        const qreal ratioLimit = getRatio();
+        if (m_initialValues.ratio != ratioLimit)
+            torrent->setRatioLimit(ratioLimit);
+
+        const int seedingTimeLimit = getSeedingTime();
+        if (m_initialValues.seedingTime != seedingTimeLimit)
+            torrent->setSeedingTimeLimit(seedingTimeLimit);
+
+        if (!torrent->isPrivate())
+        {
+            if (m_initialValues.disableDHT != m_ui->checkDisableDHT->checkState())
+                torrent->setDHTDisabled(m_ui->checkDisableDHT->isChecked());
+            if (m_initialValues.disablePEX != m_ui->checkDisablePEX->checkState())
+                torrent->setPEXDisabled(m_ui->checkDisablePEX->isChecked());
+            if (m_initialValues.disableLSD != m_ui->checkDisableLSD->checkState())
+                torrent->setLSDDisabled(m_ui->checkDisableLSD->isChecked());
+        }
+    }
+
+    QDialog::accept();
+}
+
+qreal TorrentOptionsDialog::getRatio() const
+{
+    if (m_ui->buttonGroup->checkedId() == -1) // No radio button is selected
+        return -9; // invalid number to signify different torrents ratios
```

Comment:
I would declare constant MIXED_RATIOS to indicate it instead of using numbers directly.

Reply:
I declared MIXED_SHARE_LIMITS that is used for both ratio and seeding time. I can change it to 2 constants if that's confusing.